sympy/printing/smtlib.py

```diff
@@ -0,0 +1,515 @@
+        assert len(type_signature) > 0
+        params_signature = f"({' '.join(type_signature[:-1])})"
+        return_signature = type_signature[-1]
+        return p._s_expr('declare-fun', [type(sym), params_signature, return_signature])
+
+    else:
+        log_warn(f"Non-Symbol/Function `{sym}` will not be declared.")
+        return None
+
+
+def _auto_assert_smtlib(e: Expr, p: SMTLibPrinter, log_warn=print):
+    if e.is_Boolean or e.is_Relational or (
+        e in p.symbol_table and p.symbol_table[e] == bool
+    ) or (
+        e.is_Function and
+        type(e) in p.symbol_table and
+        p.symbol_table[type(e)].__args__[-1] == bool
+    ):
+        return p._s_expr('assert', [e])
+    else:
+        log_warn(f"Non-Boolean expression `{e}` will not be asserted. Converting to SMTLib verbatim.")
+        return e
+
+
+def _auto_infer_smtlib_types(
+    *exprs: Basic,
+    symbol_table: dict = None,
+    log_warn=print
+) -> dict:
+    # [TYPE INFERENCE RULES]
+    # X is alone in an expr => X is bool
+    # X in BooleanFunction.args => X is bool
+    # X matches to a bool param of a symbol_table function => X is bool
+    # X matches to an int param of a symbol_table function => X is int
+    # X.is_integer => X is int
+    # X == Y, where X is T => Y is T
+
+    # [FALLBACK RULES]
+    # see _auto_declare_smtlib(..)
+    # X is not bool and X is not int and X is Symbol => X is float
+    # else (e.g. X is Function) => error. must be specified explicitly.
+
+    _symbols = dict(symbol_table) if symbol_table else {}
+
+    def safe_update(syms: set, inf):
+        for s in syms:
+            assert s.is_Symbol
+            if (old_type := _symbols.setdefault(s, inf)) != inf:
+                raise TypeError(f"Could not infer type of `{s}`. Apparently both `{old_type}` and `{inf}`?")
+
+    # EXPLICIT TYPES
+    safe_update({
+        e
+        for e in exprs
+        if e.is_Symbol
+    }, bool)
+
+    safe_update({
+        symbol
+        for e in exprs
+        for boolfunc in e.atoms(BooleanFunction)
+        for symbol in boolfunc.args
+        if symbol.is_Symbol
+    }, bool)
+
+    safe_update({
+        symbol
+        for e in exprs
+        for boolfunc in e.atoms(Function)
+        if type(boolfunc) in _symbols
+        for symbol, param in zip(boolfunc.args, _symbols[type(boolfunc)].__args__)
+        if symbol.is_Symbol and param == bool
+    }, bool)
+
+    safe_update({
+        symbol
+        for e in exprs
+        for intfunc in e.atoms(Function)
+        if type(intfunc) in _symbols
+        for symbol, param in zip(intfunc.args, _symbols[type(intfunc)].__args__)
+        if symbol.is_Symbol and param == int
+    }, int)
+
+    safe_update({
+        symbol
+        for e in exprs
+        for symbol in e.atoms(Symbol)
+        if symbol.is_integer
+    }, int)
+
+    safe_update({
+        symbol
+        for e in exprs
+        for symbol in e.atoms(Symbol)
+        if symbol.is_real and not symbol.is_integer
```

Comment:
Why `not symbol.is_integer`? Integers can also be floating point values. It is rare for a symbol to be defined with `integer=False`.

Reply:
`not symbol.is_integer` is necessary because not all reals are integers and we want to use the most specific type possible. The Real v.s. Integer distinction is very important in SMT since they use different theories/mechanisms in the solver software. Finding an integer solution to a formula is generally harder than finding a real solution.

Also note, the `float` here doesn't affect behavior, it could just as well be an enum.

Finally, just so we're on the same page, it is rare for a symbol to be explicitly defined with `integer=False`, but `integer=False` is still the default when left unspecified?

Reply:
No, the default is `integer=None`. `integer=False` means specifically it's not an integer. None in the assumptions means it is unknown/unspecified, i.e., it could be an integer or it could not be.

Reply:
oops. Python truthiness can be misleading 😅
That is still fine though, my type inference rules are met. Unless we know a number is an integer, it's real.